File: typescript/fsr-server/src/providers/polymer.ts
import { utils } from 'ethers';
import { z } from 'zod';

import { EvmLog, encodeEvmLog } from '../directives/evm_log.js';
import { DirectiveType, MAGIC_NUMBER } from '../directives/types.js';
import { FSRResponse } from '../index.js';

/**
 * Schema for Polymer directive
 * Format: [MAGIC_NUMBER,[DIRECTIVE_TYPE,[CHAIN_ID,BLOCK_NUMBER,TX_INDEX,LOG_INDEX]]
 *
 * Byte layout:
 * - MAGIC_NUMBER (32 bytes)
 * - DIRECTIVE_TYPE (1 byte)
 * - CHAIN_ID (8 bytes, u64)
 * - BLOCK_NUMBER (8 bytes, u64)
 * - TX_INDEX (4 bytes, u32)
 * - LOG_INDEX (4 bytes, u32)
 */
const PolymerDirectiveSchema = z.object({
  chainId: z.number(),
  blockNumber: z.number(),
  txIndex: z.number(),
  logIndex: z.number(),
});

type PolymerDirective = z.infer<typeof PolymerDirectiveSchema>;

export class PolymerProvider {
  private apiToken: string;
  private apiEndpoint: string;
  private maxRetries: number;

  constructor(apiToken: string, apiEndpoint: string, maxRetries: number = 5) {
    this.apiToken = apiToken;
    this.apiEndpoint = apiEndpoint;
    this.maxRetries = maxRetries;
  }

  async process(directiveHex: string): Promise<FSRResponse> {
    try {
      // Parse the directive hex string into a PolymerDirective
      const directive = this.parseDirective(directiveHex);

      // Create the proof request
      const request = {
        jsonrpc: '2.0',
        id: 1,
        method: 'log_requestProof',
        params: [
          directive.chainId,
          directive.blockNumber,
          directive.txIndex,
          directive.logIndex,
        ],
      };

      // Request the proof
      const response = await fetch(this.apiEndpoint, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Authorization: `Bearer ${this.apiToken}`,
        },
        body: JSON.stringify(request),
      });

      const { result: jobId } = await response.json();

      // Poll for the proof
      let attempts = 0;
      while (attempts < this.maxRetries) {
        const pollResponse = await fetch(this.apiEndpoint, {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            jsonrpc: '2.0',
            id: 1,
            method: 'log_queryProof',
            params: [jobId],
          }),
        });

        const { result } = await pollResponse.json();

        if (result.status === 'ready' || result.status === 'complete') {
          // Parse the proof and encode it as an EvmLogDirective
          const encodedDirective = this.parseProof(result.proof);

          return {
            result: encodedDirective,
            proof: result.proof,
          };
        }

        attempts++;
        await new Promise((resolve) => setTimeout(resolve, 2000)); // 2 second delay
      }

      throw new Error('Timeout waiting for proof');
    } catch (error) {
      if (error instanceof Error) {
        throw new Error(
          `Failed to process Polymer directive: ${error.message}`,
        );
      }
      throw new Error('Failed to process Polymer directive: Unknown error');
    }
  }

  /**
   * Proof byte layout
   *  *--------------------------------------------------*
   *  |  state root (32 bytes)                           | 0:32
   *  *--------------------------------------------------*
   *  |  signature (65 bytes)                            | 32:97
   *  *--------------------------------------------------*
   *  |  source chain ID (big endian, 4 bytes)           | 97:101
   *  *--------------------------------------------------*
   *  |  peptide height (big endian, 8 bytes)            | 101:109
   *  *--------------------------------------------------*
   *  |  source chain block height (big endian, 8 bytes) | 109:117
   *  *--------------------------------------------------*
   *  |  receipt index (big endian, 2 bytes)             | 117:119
   *  *--------------------------------------------------*
   *  |  event index (1 byte)                            | 119
   *  *--------------------------------------------------*
   *  |  number of topics (1 byte)                       | 120
   *  *--------------------------------------------------*
   *  |  event data end (big endian, 2 bytes)            | 121:123
   *  *--------------------------------------------------*
   *  |  event emitter (contract address) (20 bytes)     | 123:143
   *  *--------------------------------------------------*
   *  |  topics (32 bytes * number of topics)            | 143 + 32 * number of topics: eventDatEnd
   *  *--------------------------------------------------*
   *  |  event data (x bytes)                            | eventDataEnd:
   *  *--------------------------------------------------*
   *  |  iavl proof (x bytes)                            |
   *  *--------------------------------------------------*
   */
  private parseProof(proofHex: string): string {
    // Remove '0x' prefix if present
    const hex = proofHex.startsWith('0x') ? proofHex.slice(2) : proofHex;
    const bytes = Buffer.from(hex, 'hex');

    // Parse source chain ID (4 bytes, big endian)
    const sourceChainId = bytes.readUInt32BE(97).toString();

    // Parse event emitter (20 bytes)
    const eventEmitter = '0x' + bytes.subarray(123, 143).toString('hex');

    // Parse number of topics (1 byte)
    const numTopics = bytes[120];

    // Parse topics (32 bytes each)
    const topics: string[] = [];
    let offset = 143;
    for (let i = 0; i < numTopics; i++) {
      const topic = '0x' + bytes.subarray(offset, offset + 32).toString('hex');
      topics.push(topic);
      offset += 32;
    }

    // Parse event data end (2 bytes, big endian)
    const eventDataEnd = bytes.readUInt16BE(121);

    // Parse event data (from eventDataEnd to the end of the proof)
    const eventData = '0x' + bytes.subarray(eventDataEnd).toString('hex');

    // Create EvmLog
    const log: EvmLog = {
      chainId: sourceChainId,
      contract: eventEmitter,
      indexed: topics,
      unindexed: eventData,
    };

    // Encode the evm log
    return encodeEvmLog(log);
  }

  private parseDirective(directiveHex: string): PolymerDirective {
    // Remove '0x' prefix if present
    const hex = directiveHex.startsWith('0x')
      ? directiveHex.slice(2)
      : directiveHex;

    // Convert hex to bytes
    const bytes = Buffer.from(hex, 'hex');

    // Check magic number
    const magicNumberBytes = Buffer.from(MAGIC_NUMBER.slice(2), 'hex');
    if (bytes.length < magicNumberBytes.length + 1 || bytes[0] !== 0x5b) {
      // '['
      throw new Error('Invalid directive format: missing opening bracket');
    }

    if (
      bytes.subarray(1, magicNumberBytes.length + 1).toString('hex') !==
      magicNumberBytes.toString('hex')
    ) {
      throw new Error('Invalid magic number');
    }

    // Skip magic number and opening brackets
    let offset = magicNumberBytes.length + 2; // +2 for '[' and ','

    // Check directive type
    const directiveType = bytes[offset];
    if (directiveType !== DirectiveType.EVMLog) {
      throw new Error('Unsupported directive type');
    }
    offset += 2; // Skip directive type and comma

    // Check for opening bracket of args
    if (bytes[offset] !== 0x5b) {
      // '['
      throw new Error('Invalid args format');
    }
    offset += 1;

    // Parse args (chain_id, block_number, tx_index, log_index)
    const chainId = bytes.readBigUInt64BE(offset);
    offset += 8;
    const blockNumber = bytes.readBigUInt64BE(offset);
    offset += 8;
    const txIndex = bytes.readUInt32BE(offset);
    offset += 4;
    const logIndex = bytes.readUInt32BE(offset);

    return {
      chainId: Number(chainId),
      blockNumber: Number(blockNumber),
      txIndex,
      logIndex,
    };
  }
}

File: rust/main/agents/relayer/src/msg/metadata/polymer/polymer.rs
use base64::{engine::general_purpose, Engine};
use derive_new::new;
use ethers::types::Bytes;
use eyre::{Result, WrapErr};
use tracing::debug;

/// Message sent to the Polymer proof provider
#[derive(Debug)]
pub struct PolymerProofRequest {
    /// The chain ID of the source chain
    pub chain_id: u64,
    /// The block number of the message
    pub block_number: u64,
    /// The transaction index of the message
    pub tx_index: u32,
    /// The log index of the message
    pub log_index: u32,
}

/// Response from the Polymer proof provider
#[derive(Debug)]
pub struct PolymerProofResponse {
    /// The proof for the message
    pub proof: Bytes,
}

/// Fetches proofs for messages using Polymer's proof service
#[derive(Debug, new, Clone)]
pub struct PolymerProofProvider {
    #[new(default)]
    /// The API token for Polymer
    api_token: String,
    #[new(default)]
    /// The API endpoint for Polymer
    api_endpoint: String,
    #[new(value = "5")]
    /// The maximum number of retries for the proof request
    max_retries: u32,
}

impl Default for PolymerProofProvider {
    fn default() -> Self {
        Self {
            api_token: "b28a24ba-0cb7-4df0-ac9c-8e1430bc1360".to_string(),
            api_endpoint: "https://proof.polymer.zone".to_string(),
            max_retries: 5,
        }
    }
}

impl PolymerProofProvider {
    /// Fetch a proof for a message
    pub async fn fetch_proof(&self, request: &PolymerProofRequest) -> Result<PolymerProofResponse> {
        tracing::info!(
            chain_id = request.chain_id,
            block_number = request.block_number,
            tx_index = request.tx_index,
            log_index = request.log_index,
            "Fetching proof from Polymer proof service"
        );

        // Create the proof API client
        let client = reqwest::Client::new();

        // Request the proof
        let response = client
            .post(&self.api_endpoint)
            .header("Authorization", format!("Bearer {}", self.api_token))
            .json(&serde_json::json!({
                "jsonrpc": "2.0",
                "id": 1,
                "method": "log_requestProof",
                "params": [
                    request.chain_id,
                    request.block_number,
                    request.tx_index,
                    request.log_index
                ]
            }))
            .send()
            .await
            .wrap_err("log_requestProof")?;

        let response_json: serde_json::Value = response
            .json()
            .await
            .wrap_err("Failed to parse JSON response for log_requestProof")?;

        let job_id = response_json["result"].as_i64().ok_or_else(|| {
            eyre::eyre!("'result' field is missing or not an i64 in log_requestProof response")
        })?;

        tracing::info!(job_id, "Proof requested");

        // Poll for the proof
        let mut attempts = 0;
        loop {
            let response = client
                .post(&self.api_endpoint)
                .json(&serde_json::json!({
                    "jsonrpc": "2.0",
                    "id": 1,
                    "method": "log_queryProof",
                    "params": [job_id]
                }))
                .send()
                .await
                .wrap_err("log_queryProof")?;

            let response_json: serde_json::Value = response
                .json()
                .await
                .wrap_err("Failed to parse JSON response for log_queryProof")?;

            // Extract the nested "result" object
            let result = response_json["result"].as_object().ok_or_else(|| {
                eyre::eyre!("'result' field is missing or not an object in log_queryProof response")
            })?;

            tracing::info!(
                job_id,
                status = ?result["status"],
                "Received proof status response",
            );

            if result["status"] == "ready" || result["status"] == "complete" {
                let proof_str = result["proof"]
                    .as_str()
                    .ok_or_else(|| eyre::eyre!("'proof' field is missing or not a string"))?;
                let proof = general_purpose::STANDARD
                    .decode(proof_str)
                    .wrap_err("Failed to decode base64 proof")?;
                return Ok(PolymerProofResponse {
                    proof: Bytes::from(proof),
                });
            }

            attempts += 1;
            if attempts > self.max_retries {
                return Err(eyre::eyre!("Timeout waiting for proof"));
            }

            tracing::info!(
                job_id,
                attempt = attempts,
                max_retries = self.max_retries,
                "Polling for proof status"
            );

            // TODO: Use exponential backoff.
            tokio::time::sleep(tokio::time::Duration::from_secs(2)).await;
        }
    }
}
